File: problems/python/MeetingScheduler/src/models/meeting_room.py
from threading import Lock

from models.interval import Interval


class MeetingRoom:
    def __init__(self, room_id: int, capacity: int) -> None:
        self.id = room_id
        self.capacity = capacity
        self._blocked_intervals: list[Interval] = []
        self._lock = Lock()

    def check_and_book_interval(self, requested_capacity: int, interval: Interval) -> bool:
        with self._lock:
            if self._is_available(requested_capacity, interval):
                self._blocked_intervals.append(interval)
                self._blocked_intervals.sort(key=lambda value: (value.start_time, value.end_time))
                return True
            return False

    def _is_available(self, requested_capacity: int, interval: Interval) -> bool:
        if requested_capacity > self.capacity:
            return False

        for current in self._blocked_intervals:
            if current.start_time >= interval.end_time:
                break
            if current.is_overlap(interval):
                return False
        return True

    def release_interval(self, interval: Interval) -> None:
        with self._lock:
            for idx, current in enumerate(self._blocked_intervals):
                if current == interval:
                    del self._blocked_intervals[idx]
                    break

    def get_blocked_intervals(self) -> list[Interval]:
        with self._lock:
            return list(self._blocked_intervals)

File: problems/java/MeetingScheduler/src/models/MeetingRoom.java
package models;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.TreeSet;

public class MeetingRoom {

    private final int id;
    private final int capacity;
    private final TreeSet<Interval> blockedIntervals;

    public MeetingRoom(int id, int capacity) {
        this.id = id;
        this.capacity = capacity;
        blockedIntervals = new TreeSet<>(
                Comparator.comparing(Interval::getStartTime).thenComparing(Interval::getEndTime));
    }

    public synchronized boolean checkAndBookInterval(int requestedCapacity, Interval interval) {
        if (isAvailable(requestedCapacity, interval)) {
            blockedIntervals.add(interval);
            return true;
        }
        return false;
    }

    private boolean isAvailable(int requestedCapacity, Interval interval) {
        if (requestedCapacity > capacity) {
            return false;
        }

        for (Interval current : blockedIntervals) {
            if (!current.getStartTime().isBefore(interval.getEndTime())) {
                break;
            }
            if (current.isOverlap(interval)) {
                return false;
            }
        }
        return true;
    }

    public synchronized void releaseInterval(Interval interval) {
        blockedIntervals.remove(interval);
    }

    public List<Interval> getBlockedIntervals() {
        return new ArrayList<>(blockedIntervals);
    }

    public int getId() {
        return id;
    }

    public int getCapacity() {
        return capacity;
    }
}
